agent/api/task/task.go

```diff
@@ -709,6 +734,92 @@ func (task *Task) getAllASMSecretRequirements() map[string]apicontainer.Secret {
 	return reqs
 }
 
+// hasLogRouter returns true if the task has a log router container.
+func (task *Task) hasLogRouter() bool {
+	for _, container := range task.Containers {
+		if container.LogRouter != nil { // This is a log router container.
+			return true
+		}
+	}
+	return false
+}
+
+// initializeLogRouterResource initializes the log router task resource and adds it as a dependency of the
+// log router container.
+func (task *Task) initializeLogRouterResource(config *config.Config, resourceFields *taskresource.ResourceFields) error {
+	containerToLogOptions, err := task.collectLogRouterOptions()
+	if err != nil {
+		return errors.Wrap(err, "unable to initialize log router resource")
+	}
+
+	var logRouterResource *logrouter.LogRouterResource
+	for _, container := range task.Containers {
+		logRouter := container.LogRouter
+		if logRouter != nil {
+			var ec2InstanceID string
+			if container.Environment != nil && container.Environment[awsExecutionEnvKey] == ec2ExecutionEnv {
+				ec2InstanceID = resourceFields.EC2InstanceID
+			}
+
+			logRouterResource = logrouter.NewLogRouterResource(config.Cluster, task.Arn, task.Family+":"+task.Version,
+				ec2InstanceID, config.DataDir, logRouter.Type, logRouter.EnableECSLogMetaData, containerToLogOptions)
+			task.AddResource(logrouter.ResourceName, logRouterResource)
+			container.BuildResourceDependency(logRouterResource.GetName(), resourcestatus.ResourceCreated,
+				apicontainerstatus.ContainerCreated)
+			return nil
+		}
+	}
+
+	return errors.New("unable to initialize log router resource because there's no log router container")
```

Comment:
In order for log router set up correctly, do other containers need to wait for log router container running first before they transit to running status?

Reply:
yeah, other containers can only send logs after the log router container starts running. The design is to not add this ordering internally, but rely on customer to specify the dependency in task definition.

Reply:
Will the backend side have the logic to verify whether customers specify the dependency or not?

Reply:
no (though we probably can.. i don't have a preference on this). we added an option "fluentd-async-connect" to containers using log router (https://github.com/aws/amazon-ecs-agent/pull/2120) so that they will not fail to start if they start before the log router starts (and message will be buffered until the log router starts). If customers don't want to buffer the messages, they need to specify the container ordering. Otherwise the message will be buffered up to a limit of log sender container's memory limit before sending to log router container when it starts, if it starts later than the log sender container (see https://docs.docker.com/config/containers/logging/fluentd/, fluentd-async-connect and fluentd-buffer-limit options)